Core/clim-basic/geometry/region-predicates.lisp

```diff
@@ -1,43 +1,65 @@
 (in-package #:climi)
 
+;;; This file contains implementation of the predicate protocol. To
+;;; avoid circular dependencies and other issues we set a few rules:
+;;;
+;;; - don't use functions which methods are defined later in the file
+;;; - using bounding rectangle and transformation protocol is fine
+;;; - using composition protocol is not allowed
+;;; - use coordinate functions for comparisons
+;;; - when possible provide a method for a protocol class, and when
+;;;   beneficial provide (also) a method for a standard class
+;;; - all "region" arguments must be specialized
+;;;
+;;; Regions which must be handled:
+;;;
+;;;   - 0 dimensions: nowhere-region, point
+;;;   - 1 dimensions: polyline, elliptical-arc
+;;;   - 2 dimensions: everywhere-reigon, polygon, ellipse
+;;;   - region-set: standard-region-union, standard-region-difference,
+;;;     standard-region-intersection, standard-rectangle-set
+;;;
+;;; Additionally line and rectangle should be handled (they are
+;;; frequently used special cases of the polyline and the polygon).
+;;;
+;;; Methods should be defined from the most general to the most
+;;; specific. Methods specialized on REGION, methods being a
+;;; consequence of the dimensionality rule, unbounded regions, region
+;;; sets, and finally methods specific to a particular region type.
 
-(defmethod region-contains-position-p ((self standard-point) px py)
-  (with-slots (x y) self
-    (and (coordinate= x px) (coordinate= y py))))
+(declaim (inline line-contains-point-p))
+(defun line-contains-point-p (x1 y1 x2 y2 px py)
+  (coordinate= (* (- py y1) (- x2 x1))
+               (* (- px x1) (- y2 y1))))
+
+(declaim (inline segment-contains-point-p))
+(defun segment-contains-point-p (x1 y1 x2 y2 x y)
+  (and (coordinate-in x1 x x2)
+       (coordinate-in y1 y y2)
+       (line-contains-point-p x1 y1 x2 y2 x y)))
+
+;;; CLIM "native" coordinate system is left-handed (y grows down).
+;;; Angles are specified to grow in the counter-clockwise direction
+;;; (disregarding of the coordinate system), so we need to inverse the
+;;; y coordinate in a call to atan.
```

Comment:
need to *invert* the y coordinate in *the* call to atan.

Reply:
done